adapters/openx/openx.go

```diff
@@ -0,0 +1,190 @@
+	reqExt := openxReqExt{BidderConfig: config}
+
+	for _, imp := range request.Imp {
+		if err := preprocess(&imp, &reqExt); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		validImps = append(validImps, imp)
+	}
+
+	// If all the imps were malformed, don't bother making a server call with no impressions.
+	if len(validImps) == 0 {
+		return nil, errs
+	}
+
+	request.Imp = validImps
+
+	var err error
+	request.Ext, err = json.Marshal(reqExt)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	reqJSON, err := json.Marshal(request)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     uri,
+		Body:    reqJSON,
+		Headers: headers,
+	}, errs
+}
+
+// Mutate the imp to get it ready to send to openx.
+func preprocess(imp *openrtb.Imp, reqExt *openxReqExt) error {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return err
+	}
+
+	var openxExt openrtb_ext.ExtImpOpenx
+	if err := json.Unmarshal(bidderExt.Bidder, &openxExt); err != nil {
+		return err
+	}
+
+	reqExt.DelDomain = openxExt.DelDomain
+
+	imp.TagID = openxExt.Unit
+	imp.BidFloor = openxExt.CustomFloor
+	imp.Ext = nil
+
+	if openxExt.CustomParams != nil {
+		impExt := openxImpExt{
+			CustomParams: openxExt.CustomParams,
+		}
+		var err error
+		if imp.Ext, err = json.Marshal(impExt); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (a *OpenxAdapter) MakeBids(internalRequest *openrtb.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) ([]*adapters.TypedBid, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{fmt.Errorf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode)}
+	}
+
+	var bidResp openrtb.BidResponse
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	bids := make([]*adapters.TypedBid, 0, 5)
+
+	for _, sb := range bidResp.SeatBid {
+		for i := range sb.Bid {
+			bids = append(bids, &adapters.TypedBid{
+				Bid:     &sb.Bid[i],
+				BidType: getMediaTypeForImp(sb.Bid[i].ImpID, internalRequest.Imp),
+			})
+		}
+	}
+	return bids, nil
+}
+
+// getMediaTypeForImp figures out which media type this bid is for.
+//
+// OpenX doesn't support multi-type impressions.
+// If both banner and video exist, take banner as we do not want in-banner video.
+func getMediaTypeForImp(impId string, imps []openrtb.Imp) openrtb_ext.BidType {
+	mediaType := openrtb_ext.BidTypeBanner
+	for _, imp := range imps {
+		if imp.ID == impId {
+			if imp.Banner == nil && imp.Video != nil {
+				mediaType = openrtb_ext.BidTypeVideo
+			}
+			return mediaType
```

Comment:
This assumes that the media type is always banner or video. I think you want a check in `preprocess` to verify that the imp type is one of the two, and reject it otherwise.